### ISSUE TITLE
Sometimes does not reboot on Fedora 29/30

### ISSUE THREAD

**Issue by MichaelRiss** [agross/dnf-automatic-restart#9]
TL;DR
Maybe the manual check in the script of whether the machine needs rebooting for a new kernel should be replaced with just checking the return code of tracer.

---

The long form:

Hello Alexander,

I'm back with some more Fedora machines not rebooting. I have not entirely traced the problem to the root cause yet, but the tip of the iceberg looks like this:
The current method to fetch the version number of the latest installed kernel out of `/boot/grub2/grubenv` does not work anymore (reliable). Something seems to have changed in how Fedora generates their boot menu ids. On F29 I still get on some machines the old 
```
saved_entry=Fedora (5.0.19-200.fc29.x86_64) 29 (Twenty Nine)
```
but on one machine I already get:
```
saved_entry=gnulinux-4.18.16-300.fc29.x86_64-advanced-23d71b87-e4e6-4707-9c5f-0f49b7b98dc1
```
The second boot menu id is clearly broken as it also refers to same kernel: `5.0.19`. This second machine is also older, meaning: it got installed earlier in the F29 life cycle, the first machine with the proper, old-style menu id got installed later in the F29 life cycle.

On F30 there is a completely different picture:
Now, the entries look like this: 
```
saved_entry=ee50373e510a4336a13d8e768c6f0a4e-5.1.6-300.fc30.x86_64
```
and this is both on machines that got installed early in the life cycle and on machines that got installed in the last couple of days. 
Now, I could speculate about a potential hickup early in the life cycle of the F29 that somehow mangled the then-current kernel version string into the boot menu ids of all following kernels and that this problem probably prompted a rework on how the menu ids get generated in F30, but I don't know anything for sure about this and currently I don't have the time to dive into the Fedora bugzilla ticket jungle to find out.

However in summary: Currently, the `/boot/grub2/grubenv` information source is not reliable.

And then I saw your comment:
```
# Kernel-only updates are not detected by tracer.
# https://github.com/FrostyX/tracer/issues/45
```
So, back in the days you added the custom reboot detection code because the corresponding function of tracer did not work. Since then some time has passed and I wondered whether they did make some progress on this issue and - they did: tracer can now detect whether a reboot is necessary.
Testing `tracer` on all machines that currently fail to detect the necessity to reboot, yielded the return code `104` - reboot necessary.
I'd say switching the reboot detection to tracer seems to be the easiest way to solve this problem.

At the moment, I don't have the time to implement it, but I wanted to give you already an advance heads-up.

cheers
Michi

**Comment by agross** [comment]
Hi Michi and @maddanio,

Thank you for reporting in such detail and doing the research! What do you think about #10? 

Can you please test the changes on your systems?

Thanks,

Alex

**Comment by MichaelRiss** [comment]
Hello Alex,

it's looking good, my machines reboot correctly now. 
Thanks for the quick fix!

Cheers,
Michi
